src/elements/core/mixins/form-associated-radio-button-mixin.ts

```diff
@@ -0,0 +1,314 @@
+      if (this.checked) {
+        this.internals.setFormValue(this.value);
+      }
+    }
+
+    /**
+     * Called on `checked` change
+     * If 'checked', update the value. Otherwise, reset it.
+     */
+    protected updateFormValueOnCheckedChange(): void {
+      this.internals.setFormValue(this.checked ? this.value : null);
+    }
+
+    /**
+     * Only a single radio should be focusable in the group. Defined as:
+     * - the checked radio;
+     * - the first non-disabled radio in DOM order;
+     */
+    protected updateFocusableRadios(): void {
+      if (!this._didLoad) {
+        return;
+      }
+      const radios = this._orderedGroupedRadios();
+      const checkedIndex = radios.findIndex((r) => r.checked && !r.disabled && !r.formDisabled);
+      const focusableIndex =
+        checkedIndex !== -1
+          ? checkedIndex
+          : radios.findIndex((r) => !r.disabled && !r.formDisabled); // Get the first focusable radio
+
+      if (focusableIndex !== -1) {
+        radios[focusableIndex].tabIndex = 0;
+        radios.splice(focusableIndex, 1);
+      }
+
+      // Reset tabIndex on other radios
+      radios.forEach((r) => r.removeAttribute('tabindex'));
+    }
+
+    protected async navigateByKeyboard(next: SbbFormAssociatedRadioButtonElement): Promise<void> {
+      next.checked = true;
+      this.emitChangeEvents();
+
+      await next.updateComplete;
+      next.focus();
+    }
+
+    protected emitChangeEvents(): void {
+      // Manually dispatch events to simulate a user interaction
+      this.dispatchEvent(
+        new InputEvent('input', { bubbles: true, cancelable: true, composed: true }),
+      );
+      this.dispatchEvent(new Event('change', { bubbles: true }));
+    }
+
+    /**
+     * Add `this` to the radioButton registry
+     */
+    private _connectToRegistry(name = this.name): void {
+      if (!name) {
+        return;
+      }
+      RadioButtonRegistry.addRadioToGroup(
+        this as unknown as SbbFormAssociatedRadioButtonMixinType,
+        name,
+      );
+    }
+
+    /**
+     * Remove `this` from the radioButton registry
+     */
+    private _disconnectFromRegistry(name = this.name): void {
+      if (!name) {
+        return;
+      }
+      RadioButtonRegistry.removeRadioFromGroup(
+        this as unknown as SbbFormAssociatedRadioButtonMixinType,
+        name,
+      );
+    }
+
+    /**
+     * Deselect other radio of the same group
+     */
+    private _deselectGroupedRadios(): void {
+      RadioButtonRegistry.getRadios(this.name)
+        .filter((r) => r !== (this as unknown as SbbFormAssociatedRadioButtonMixinType))
+        .forEach((r) => (r.checked = false));
+    }
+
+    /**
+     * Return the grouped radios in DOM order
+     */
+    private _orderedGroupedRadios(groupName = this.name): SbbFormAssociatedRadioButtonElement[] {
+      return Array.from(
+        (this.form ?? document).querySelectorAll<SbbFormAssociatedRadioButtonElement>(
```

Comment:
Because the registry does not ensure that the radios are in the correct DOM order (which is needed both for keyboard navigation and to make the first radio focusable).
And the only way (I've found) to know the DOM order is to querySelect.

Reply:
Hmm I see. I feel that this solution is not that stable, as e.g. if you have an app using shadow DOM (e.g. our visual regression app). this kind of selection via document probably doesn't work. But the registry would be shadow DOM safe. Would it maybe be possible to find out the DOM order by using https://developer.mozilla.org/en-US/docs/Web/API/Node/compareDocumentPosition? (sort by this comparator)